tests/test_environments.py

```diff
@@ -276,3 +341,340 @@ def test_text_environment_run(self, mock_generate):
+        self.assertTrue(torch.all(combined_cache[0][1] == expected_cache[0][1]))
+        self.assertEqual(len(combined_cache[1]), len(expected_cache[1]))
+        self.assertTrue(torch.all(combined_cache[1][0] == expected_cache[1][0]))
+        self.assertTrue(torch.all(combined_cache[1][1] == expected_cache[1][1]))
+        self.assertTrue(torch.all(combined_attention_masks == expected_attention_mask))
+        self.assertTrue(torch.all(combined_input_ids == expected_input_ids))
+
+    def test_get_batched_cache(self):
+        env = TextEnvironment(
+            self.model,
+            self.tokenizer,
+            tools={"DummyTool": DummyTool()},
+            reward_fn=lambda x: [torch.tensor(i) for i, _ in enumerate(x)],
+            prompt="I am a prompt!\n",
+            max_turns=2,
+        )
+
+        cache = reshape_cache(
+            (
+                (torch.tensor([[1], [2], [3]]), torch.tensor([[4], [5], [6]])),
+                (torch.tensor([[7], [8], [9]]), torch.tensor([[10], [11], [12]])),
+            )
+        )
+        attention_masks = torch.tensor([[1], [2], [3]])
+        input_ids = torch.tensor([[4], [5], [6]])
+        batched_cache, batched_attention_masks, batched_input_ids = env._get_batched_cache(
+            1, 3, cache, attention_masks, input_ids
+        )
+        expected_cache = reshape_cache(
+            (
+                (torch.tensor([[2], [3]]), torch.tensor([[5], [6]])),
+                (torch.tensor([[8], [9]]), torch.tensor([[11], [12]])),
+            )
+        )
+
+        self.assertEqual(len(batched_cache), len(expected_cache))
+        self.assertEqual(len(batched_cache[0]), len(expected_cache[0]))
+        self.assertTrue(torch.all(batched_cache[0][0] == expected_cache[0][0]))
+        self.assertTrue(torch.all(batched_cache[0][1] == expected_cache[0][1]))
+        self.assertEqual(len(batched_cache[1]), len(expected_cache[1]))
+        self.assertTrue(torch.all(batched_cache[1][0] == expected_cache[1][0]))
+        self.assertTrue(torch.all(batched_cache[1][1] == expected_cache[1][1]))
+
+        expected_attention_mask = torch.tensor([[2], [3]])
+        self.assertTrue(torch.all(batched_attention_masks == expected_attention_mask))
+
+        expected_input_ids = torch.tensor([[5], [6]])
+        self.assertTrue(torch.all(batched_input_ids == expected_input_ids))
+
+    def test_same_is_none(self):
+        env = TextEnvironment(
+            self.model,
+            self.tokenizer,
+            tools=[DummyTool()],
+            reward_fn=lambda x: torch.tensor(1),
+            prompt="I am a prompt!\n",
+        )
+        self.assertTrue(env._same_is_none("", ""))
+        self.assertTrue(env._same_is_none(None, None))
+        self.assertFalse(env._same_is_none("", None))
+        self.assertFalse(env._same_is_none(None, ""))
+        self.assertTrue(env._same_is_none(None))
+        self.assertTrue(env._same_is_none(""))
```

Comment:
We usually don't test private methods

Reply:
Do you want me to remove a test, that's already written? I can of course, although for the other private methods I think the tests are sensible.